blockchain/tx_list.go

```diff
@@ -308,30 +308,35 @@ func (l *txList) Overlaps(tx *types.Transaction) bool {
 //
 // If the new transaction is accepted into the list, the lists' cost and gas
 // thresholds are also potentially updated.
-func (l *txList) Add(tx *types.Transaction, priceBump uint64) (bool, *types.Transaction) {
+func (l *txList) Add(tx *types.Transaction, priceBump uint64, kip71Hardforked bool) (bool, *types.Transaction) {
 	// If there's an older better transaction, abort
 	old := l.txs.Get(tx.Nonce())
 	if old != nil {
-		// TODO : Need to KIP-71 hardfork.
-		//if tx.Type().IsCancelTransaction() {
-		//	logger.Trace("New tx is a cancel transaction. replace it!", "old", old.String(), "new", tx.String())
-		//} else {
-		//	logger.Trace("already nonce exist", "nonce", tx.Nonce(), "with gasprice", old.GasPrice(), "priceBump", priceBump, "new tx.gasprice", tx.GasPrice())
-		//	return false, nil
-		//}
-
-		// If gas price of older is bigger than newer, abort.
-		if old.GasPrice().Cmp(tx.GasPrice()) >= 0 {
-			logger.Trace("already nonce exist", "nonce", tx.Nonce(), "with gasprice", old.GasPrice(), "priceBump", priceBump, "new tx.gasprice", tx.GasPrice())
-			return false, nil
+		if kip71Hardforked {
+			// If tx is CancelTransaction, replace it even thought tx has lower gasPrice than previous tx.
+			if tx.Type().IsCancelTransaction() {
+				logger.Trace("New tx is a cancel transaction. replace it!", "old", old.String(), "new", tx.String())
+			}
+
+			// If gas price of older is bigger than newer, abort.
+			if old.GasPrice().Cmp(tx.GasPrice()) >= 0 {
```

Comment:
I think this should be `else if` now to connect with above if.

Reply:
Thanks! It is fixed.